remoulade/state/backends/postgres.py

```diff
@@ -0,0 +1,159 @@
+
+from pytz import utc
+from sqlalchemy import Column, DateTime, Float, LargeBinary, SmallInteger, String, create_engine, inspect, text
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm.session import sessionmaker
+
+from remoulade import Encoder
+from remoulade.state import State, StateBackend
+
+Base = declarative_base()
+
+DEFAULT_POSTGRES_URI = "postgresql://remoulade@localhost:5432/remoulade"
+DB_VERSION = 1
+T = TypeVar("T", bound="StoredState")
+
+
+class StoredState(Base):
+
+    __tablename__ = "states"
+
+    message_id = Column(String(length=36), primary_key=True, index=True)
+    status = Column(String(length=10), index=True)
+    actor_name = Column(String(length=79), index=True)
+    args = Column(LargeBinary)
+    kwargs = Column(LargeBinary)
+    options = Column(LargeBinary)
+    priority = Column(SmallInteger)
+    progress = Column(Float)
+    enqueued_datetime = Column(DateTime(timezone=True), index=True)
+    started_datetime = Column(DateTime(timezone=True), index=True)
+    end_datetime = Column(DateTime(timezone=True), index=True)
+    group_id = Column(String(length=36))
+
+    def as_state(self, encoder: Encoder) -> State:
+        state_dict = {}
+        mapper = inspect(StoredState)
+        for column in mapper.attrs:
+            column_value = getattr(self, column.key)
+            if column_value is None:
+                continue
+            if column.key in ["args", "kwargs", "options"]:
+                column_value = encoder.decode(column_value)
+            state_dict[column.key] = column_value
+        return State.from_dict(state_dict)
+
+    @classmethod
+    def from_state(cls: Type[T], state: State, max_size: int, encoder: Encoder) -> T:
+        state_dict = state.as_dict()
+        for key in ["args", "kwargs", "options"]:
+            if key in state_dict:
+                encoded_value = encoder.encode(state_dict[key])
+                state_dict[key] = encoded_value if sys.getsizeof(encoded_value) <= max_size else None
+        return cls(**state_dict)
+
+
+class StateVersion(Base):
+
+    __tablename__ = "version"
+
+    version = Column(SmallInteger, primary_key=True)
+
+
+class PostgresBackend(StateBackend):
+    def __init__(
+        self,
+        *,
+        namespace: str = "remoulade-state",
+        encoder: Optional[Encoder] = None,
+        client: Optional[sessionmaker] = None,
+        max_size: int = 2000000,
+        url: Optional[str] = None,
+    ):
+        self.url = url or DEFAULT_POSTGRES_URI
+        super().__init__(namespace=namespace, encoder=encoder)
+        self.client = client or sessionmaker(create_engine(self.url))
+        self.init_db()
+        self.max_size = max_size
+        self.lock = threading.Lock()
+
+    def init_db(self):
+        with self.client.begin() as session:
+            bind = session.get_bind()
+            insp = inspect(bind)
+            version_exists = insp.has_table("version")
```

Comment:
I found the logic a bit complicated and I am not sure the create all works when you already have a table.
What do you think of something like this?
```python
if not insp.has_table("version"):
    Base.metadata.create_all(bind=bind, tables=[StateVersion.__table__])

state_version = session.query(StateVersion).first()    ​
if state_version is None:
  ​session.add(StateVersion(version=DB_VERSION))

if not insp.has_table("states"):
    Base.metadata.create_all(bind=bind, tables=[StoredState.__table__])
elif state_version != DB_VERSION:
    StoredState.__table__.drop(bind=bind)
    Base.metadata.create_all(bind=bind, tables=[StoredState.__table__])
    state_version.version = DB_VERSION
```

Reply:
ok I saw that you test it well, so it is more about readability